core/src/main/java/org/infinispan/expiration/impl/ClusterExpirationManager.java

```diff
@@ -219,7 +229,14 @@ private boolean pollForCompletion(BlockingQueue<CompletableFuture<?>> queue, lon
             log.tracef("Submitting expiration removal for key %s which had lifespan of %s", toStr(key), lifespan);
          }
          AdvancedCache<K, V> cacheToUse = skipLocking ? cache.withFlags(Flag.SKIP_LOCKING) : cache;
-         CompletableFuture<Void> future = cacheToUse.removeLifespanExpired(key, value, lifespan);
+         CompletableFuture<Void> future;
+         if (transactional) {
+            // Transactional is still blocking - to be removed later
```

Comment:
IMO we could make `cache.removeLifespanExpired()` submit a task to an executor internally when it's not part of a transaction, but the only way to make it realy non-blocking is to avoid any interaction with JTA, at which point we might as well remove it from the `AdvancedCache` API.

Reply:
Yeah, I think we may have to remove it from the API to more cleanly fix this in the future. This is mostly just a stop gap until that can be done.